### ISSUE TITLE
DROID 6.1.5 boundary issues...

### ISSUE THREAD

**Issue by ross-spencer** [digital-preservation/droid#85]
I'm working on a 32-bit and 64-bit windows portable exe signature and seem to have found an issue where DROID is ignoring the boundaries of a signature if there is a positive match somewhere else in the file. 

Attached is a ZIP containing the signature file and sample EXE skeleton file which demonstrates the problem. I've also extracted the signature file XML below.

The 32-bit signature should match at offset 264 dec. 
There seems to be a multiple identification of my proposed 64-bit signature at 129,000 dec. 

There may be something wrong with the signature syntax but I'm matching approximately 1,680 other random executable files from my machine without difficulty. Having double and triple checked, I'm not sure I can spot anything else that might be wrong with the sig. 

The sequences below approximate to: 

```
4D5A{126-128500}50450000{20}0B02{66}[0000:1000] 
```

and: 

```
4D5A{126-128500}50450000{20}0B01{66}[0000:1000]
```

Signature file:

```
  <?xml version="1.0" encoding="UTF-8"?>
  <FFSignatureFile xmlns="http://www.nationalarchives.gov.uk/pronom/SignatureFile" Version="1" DateCreated="2016-02-16T03:15:29+00:00">
    <InternalSignatureCollection>
    <InternalSignature ID="1" Specificity="Specific">
    <ByteSequence Reference="BOFoffset">
      <SubSequence MinFragLength="0" Position="1" SubSeqMaxOffset="0" SubSeqMinOffset="0">
        <Sequence>4D5A</Sequence>
        <DefaultShift>3</DefaultShift>
        <Shift Byte="4D">2</Shift>
        <Shift Byte="5A">1</Shift>
      </SubSequence>
      <SubSequence MinFragLength="0" Position="2" SubSeqMinOffset="126" SubSeqMaxOffset="128500">
        <Sequence>50450000</Sequence>
        <DefaultShift>5</DefaultShift>
        <Shift Byte="50">4</Shift>
        <Shift Byte="45">3</Shift>
        <Shift Byte="00">1</Shift>
        <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B02</RightFragment>
        <RightFragment MaxOffset="66" MinOffset="66" Position="2">[0000:1000]</RightFragment>
      </SubSequence>
    </ByteSequence>
  </InternalSignature>
    <InternalSignature ID="2" Specificity="Specific">
    <ByteSequence Reference="BOFoffset">
      <SubSequence MinFragLength="0" Position="1" SubSeqMaxOffset="0" SubSeqMinOffset="0">
        <Sequence>4D5A</Sequence>
        <DefaultShift>3</DefaultShift>
        <Shift Byte="4D">2</Shift>
        <Shift Byte="5A">1</Shift>
      </SubSequence>
      <SubSequence MinFragLength="0" Position="2" SubSeqMinOffset="126" SubSeqMaxOffset="128500">
        <Sequence>50450000</Sequence>
        <DefaultShift>5</DefaultShift>
        <Shift Byte="50">4</Shift>
        <Shift Byte="45">3</Shift>
        <Shift Byte="00">1</Shift>
        <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B01</RightFragment>
        <RightFragment MaxOffset="66" MinOffset="66" Position="2">[0000:1000]</RightFragment>
      </SubSequence>
    </ByteSequence>
  </InternalSignature>
  </InternalSignatureCollection>
  <FileFormatCollection>
    <FileFormat ID="1" Name="portable-exe-64-bit" PUID="exp/1" Version="" MIMEType="application/octet-stream">
      <InternalSignatureID>1</InternalSignatureID>
      <Extension>exe</Extension>
    </FileFormat>
      <FileFormat ID="2" Name="portable-exe-32-bit" PUID="exp/2" Version="" MIMEType="application/octet-stream">
      <InternalSignatureID>2</InternalSignatureID>
      <Extension>exe</Extension>
    </FileFormat>
  </FileFormatCollection></FFSignatureFile>
```

I appreciate the PRONOM stored procedure may output slightly different DROID XML.

If there are any problems with the attached ZIP, the skeleton can be copy and pasted from Gist here: https://gist.github.com/ross-spencer/c79e30cc53b6e8f0cbb3 (129748 bytes) 

![image](https://cloud.githubusercontent.com/assets/1880412/13069087/354f4408-d4e6-11e5-9c47-c8e942fa9ef5.png)

[bad-boundaries.zip](https://github.com/digital-preservation/droid/files/131976/bad-boundaries.zip)

**Comment by nishihatapalmer** [comment]
Hi Ross, 

What is the [0000:1000] supposed to match?  DROID works at a byte level.  This syntax would be interpreted as a range of bytes 0x00 to 0x10.  There are also two 0x00 bytes bookending the range in the set,  but that doesn't change the set of bytes to match. 

Regards 

Matt P.

**Comment by ross-spencer** [comment]
Hi Matt, 

Rationale here: https://groups.google.com/forum/#!topic/droid-list/QRQ9LScT8iw

Essentially matching subsystem WORD in the IMAGE_OPTIONAL_HEADER. There are 17 potential values, including zero, and so 0x00 to 0x10. The WORD is where the bookending 0x00 comes from, do you think it better to be [00:10]00 at the end of the signature? 

Thanks for looking at this,

Ross

**Comment by nishihatapalmer** [comment]
Hi Ross,

Yes I think it should be [00:10] 00

Assuming we have the right endianness...

Putting zeros inside the square brackets makes them part of the spec to
match a byte at that position, and it's redundant anyway since you already
have 00 in the range to match.

Let me know how you get on...

Cheers

Matt

**Comment by nishihatapalmer** [comment]
I seem to recall earlier versions of DROID may have been able to process 16
bit values directly.   But it certainly doesn't do this now.

I think it was stripped out as unnecessary complexity after DROID 4.  You
get into endian issues,  and I don't think it was ever used in a signature.

Could be wrong,  it was a long time ago!

Matt

**Comment by snail1966** [comment]
To note by way of background information: It is necessary to change the default DROID "maximum bytes to scan" setting, e.g. to -1 to see this issue in the sample ZIP.  i.e. with unlimited scanning, exe-boundaries.exe is identified as matching both the exp1 and exp2 sample signatures.  With the default setting it is only matched against exp/2 (the 32-bit signature).
Regards, Brian

**Comment by ross-spencer** [comment]
Thanks Brian. I should have posted my settings too. 

Matt, I just cleared all the DROID folders and started afresh with -1 and the this signature file:

      <?xml version="1.0" encoding="UTF-8"?>
      <FFSignatureFile xmlns="http://www.nationalarchives.gov.uk/pronom/SignatureFile" Version="1" DateCreated="2016-02-16T03:15:29+00:00">
        <InternalSignatureCollection>
        <InternalSignature ID="1" Specificity="Specific">
        <ByteSequence Reference="BOFoffset">
          <SubSequence MinFragLength="0" Position="1" SubSeqMaxOffset="0" SubSeqMinOffset="0">
            <Sequence>4D5A</Sequence>
            <DefaultShift>3</DefaultShift>
            <Shift Byte="4D">2</Shift>
            <Shift Byte="5A">1</Shift>
          </SubSequence>
          <SubSequence MinFragLength="0" Position="2" SubSeqMinOffset="126" SubSeqMaxOffset="128500">
            <Sequence>50450000</Sequence>
            <DefaultShift>5</DefaultShift>
            <Shift Byte="50">4</Shift>
            <Shift Byte="45">3</Shift>
            <Shift Byte="00">1</Shift>
            <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B02</RightFragment>
            <RightFragment MaxOffset="66" MinOffset="66" Position="2">[00:10]00</RightFragment>
          </SubSequence>
        </ByteSequence>
      </InternalSignature>
        <InternalSignature ID="2" Specificity="Specific">
        <ByteSequence Reference="BOFoffset">
          <SubSequence MinFragLength="0" Position="1" SubSeqMaxOffset="0" SubSeqMinOffset="0">
            <Sequence>4D5A</Sequence>
            <DefaultShift>3</DefaultShift>
            <Shift Byte="4D">2</Shift>
            <Shift Byte="5A">1</Shift>
          </SubSequence>
          <SubSequence MinFragLength="0" Position="2" SubSeqMinOffset="126" SubSeqMaxOffset="128500">
            <Sequence>50450000</Sequence>
            <DefaultShift>5</DefaultShift>
            <Shift Byte="50">4</Shift>
            <Shift Byte="45">3</Shift>
            <Shift Byte="00">1</Shift>
            <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B01</RightFragment>
            <RightFragment MaxOffset="66" MinOffset="66" Position="2">[00:10]00</RightFragment>
          </SubSequence>
        </ByteSequence>
      </InternalSignature>
      </InternalSignatureCollection>
      <FileFormatCollection>
        <FileFormat ID="1" Name="portable-exe-64-bit" PUID="exp/1" Version="" MIMEType="application/octet-stream">
          <InternalSignatureID>1</InternalSignatureID>
          <Extension>exe</Extension>
        </FileFormat>
          <FileFormat ID="2" Name="portable-exe-32-bit" PUID="exp/2" Version="" MIMEType="application/octet-stream">
          <InternalSignatureID>2</InternalSignatureID>
          <Extension>exe</Extension>
        </FileFormat>
      </FileFormatCollection></FFSignatureFile>


Unfortunately for both signature files I get identical results: 

15 multiple-ids
0 zero-id
933 64-bit portable exes
742 32-bit portable exes

**Comment by snail1966** [comment]
This is indeed an issue with DROID 6.1.5 - it identifies the test file exe-boundaries.exe  only against the 32 bit format exp/2 whose signature occurs first in the file.  With the current release (6.2.1) the test file exe-boundaries.exe  is identified against both exp/1 and exp/2 (using both versions of the test signature file as discussed above).  This is expected behaviour since both signatures occur within the file and the specified offset requirements for the sub-sequences within the file and relative to one another, and the right fragments within  the sub-sequences, are met.

It should be noted incidentally that when testing with dummy signatures in this way, there may sometimes be no identifications at all on the first profile run after installing a new signature file.  I have raised a separate issue (no. 97) regarding this point.,
Regards, Brian

**Comment by Dclipsham** [comment]
The 64 bit sequence shouldn't match as it falls outside of the relative range.
we match 0x4D5A then 128998 bytes later we hit 0x50450000 whereas the signature given allows for only up to 128500 bytes relative to 0x4D5A

That being said, DROID 6.2.1 is giving me the _desired behaviour_ of only matching against the 32 bit signature - fmt/899

The PRONOM stored procedure has indeed outputted slightly different XML:
`        <InternalSignature ID="1249" Specificity="Specific">
            <ByteSequence Reference="BOFoffset">
                <SubSequence MinFragLength="128" Position="1"
                    SubSeqMaxOffset="0" SubSeqMinOffset="0">
                    <Sequence>50450000</Sequence>
                    <DefaultShift>5</DefaultShift>
                    <Shift Byte="00">1</Shift>
                    <Shift Byte="45">3</Shift>
                    <Shift Byte="50">4</Shift>
                    <LeftFragment MaxOffset="128500" MinOffset="126" Position="1">4D5A</LeftFragment>
                    <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B01</RightFragment>
                    <RightFragment MaxOffset="66" MinOffset="66" Position="2">[0000:1000]</RightFragment>
                </SubSequence>
            </ByteSequence>
        </InternalSignature>
        <InternalSignature ID="1251" Specificity="Specific">
            <ByteSequence Reference="BOFoffset">
                <SubSequence MinFragLength="128" Position="1"
                    SubSeqMaxOffset="0" SubSeqMinOffset="0">
                    <Sequence>50450000</Sequence>
                    <DefaultShift>5</DefaultShift>
                    <Shift Byte="00">1</Shift>
                    <Shift Byte="45">3</Shift>
                    <Shift Byte="50">4</Shift>
                    <LeftFragment MaxOffset="128500" MinOffset="126" Position="1">4D5A</LeftFragment>
                    <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B02</RightFragment>
                    <RightFragment MaxOffset="66" MinOffset="66" Position="2">[0000:1000]</RightFragment>
                </SubSequence>
            </ByteSequence>
        </InternalSignature>`

N.B. I'll adjust the signatures to [00:10]00 in the next release.

Ross, can you confirm (or otherwise) that you currently get the behaviour you expect?

David

**Comment by ross-spencer** [comment]
Hi David,

I just took the time to remind myself of this issue. I do see the behavior you talk about. 

The reason for the difference in signature between PRONOM and the SD Utility at the time is because I was creating two sequences rather than using just a single regex string. If I recreate one of the signatures today we see: 


SD Utility Output: 

      <InternalSignature ID="1" Specificity="Specific">
        <ByteSequence Reference="BOFoffset">
          <SubSequence MinFragLength="128" Position="1" SubSeqMaxOffset="0" SubSeqMinOffset="0">
            <Sequence>50450000</Sequence>
            <DefaultShift>5</DefaultShift>
            <Shift Byte="50">4</Shift>
            <Shift Byte="45">3</Shift>
            <Shift Byte="00">1</Shift>
            <LeftFragment MaxOffset="128500" MinOffset="126" Position="1">4D5A</LeftFragment>
            <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B02</RightFragment>
            <RightFragment MaxOffset="66" MinOffset="66" Position="2">[0000:1000]</RightFragment>
          </SubSequence>
        </ByteSequence>
      </InternalSignature>


PRONOM output: 

      <InternalSignature ID="1251" Specificity="Specific">
        <ByteSequence Reference="BOFoffset">
            <SubSequence MinFragLength="128" Position="1"
                SubSeqMaxOffset="0" SubSeqMinOffset="0">
                <Sequence>50450000</Sequence>
                <DefaultShift>5</DefaultShift>
                <Shift Byte="00">1</Shift>
                <Shift Byte="45">3</Shift>
                <Shift Byte="50">4</Shift>
                <LeftFragment MaxOffset="128500" MinOffset="126" Position="1">4D5A</LeftFragment>
                <RightFragment MaxOffset="20" MinOffset="20" Position="1">0B02</RightFragment>
                <RightFragment MaxOffset="66" MinOffset="66" Position="2">[0000:1000]</RightFragment>
            </SubSequence>
        </ByteSequence>
      </InternalSignature>

The new sequence treating the signature as a single regular expression results in unambiguous ID of one exe type over another. 

The original sequence at the very beginning of this trail still won't work in DROID (multiple ID). I wonder if this is related to https://github.com/digital-preservation/droid/issues/105 where there are two sub-sequences that are causing a match where there shouldn't be one?

**Comment by snail1966** [comment]
I've recently revisited this issue and the situation appears to be as follows:
Using the original signature file posted above, and running DROID in "no profile" mode from the  command line, the sample file is identified against both signatures.  This behaviour is the same in the current live DROID (6.2.1), the previous version (6.1.5) and the current build with changes to allow for more accurate checking of fragment offsets (at https://github.com/snail1966/droid/tree/multiFragmentInstances).  Possibly I was using the default 65536 max bytes setting when I reported 6.1.5 GUI identifying against only the 32- bit signature - changing this to -1 results in an identification against both signatures.  This is also the case with 6.2.1 and the build mentioned above (though in all three versions you may see no identifications in the GUI due to an unrelated bug affecting initial profile runs - see #97).

The use of [0000:0010] vs [00:10] makes no difference, DROID 6.2.1 onwards (using Byteseek 2) assigns a ByteRangeMatcher and ByteMatcherSearcher with a range of 0-16 to check this fragment in either case.  Though, as discussed above, the latter syntax is more correct, given that the bookends are ignored and that DROID 6 does not support multi-byte range searches.

Using the alternative PRONOM signature (Dclipsham's post of 5th September above), the sample file is identified only against the 32 bit signature.  Again, this is the same behaviour in all 3 versions above (again verified in no profile mode - I suspect the GUI behaviour would be the same, again allowing for  issue #97, but haven't exhaustively checked with all builds/configurations).

There is actually a crucial difference between the two signatures.  The initial sample has two sub-sequences within a single byte sequence, whereas in the 2nd (PRONOM generated) sample, there are two separate byte sequences.  In the first case, DROID does indeed ignore any maximum offset specified for the 2nd subsequence.  The crucial line in the code is in ByteSequence.java, line 340:

             final boolean fullFileScan = !(subSequenceIndex == firstSequenceToMatch && isFixedStart);

Since this is not the first sub-sequence, _fullFileScan_ evaluates to true, and this means that the offset is effectively ignored when setting the end of the search window in SubSequence.java - findSequenceFromPosition.  If it evaluated to false, then the offset would be accounted for and only the 32 bit signature would be identified. This behaviour is by design - any sub-sequences after the first one are meant to follow * wildcards, and no maximum offset is applicable to the *. See also the Digital Preservation Technical Paper at http://www.nationalarchives.gov.uk/aboutapps/fileformat/pdf/automatic_format_identification.pdf, in particular Appendix 2 on pre-processing signatures. In the example here, the 2nd subsequence has no maximum offset and the minimum offset of 5 covers only the {5} at the beginning of this subsequence, not the preceding * after the first subsequence.
Regards, Brian

**Comment by paulyoung84** [comment]
Fixed in 6.3
